src/main/scala/vsys/blockchain/contract/DataType.scala

```diff
@@ -1,34 +1,162 @@
+    lenFixed          = true,
+    maxLen            = AmountLength,
+    deserializer      = Longs.fromByteArray(_),
+    serializer        = Longs.toByteArray(_),
+    jsonifier         = Json.toJson(_),
+    extValidator      = Longs.fromByteArray(_) >= 0)
+
+  val Int32           = DataTypeVal[Int](4,
+    lenFixed          = true,
+    maxLen            = 4,
+    deserializer      = Ints.fromByteArray(_),
+    serializer        = Ints.toByteArray(_),
+    jsonifier         = Json.toJson(_),
+    extValidator      = Ints.fromByteArray(_) >= 0)
+
+  val ShortText       = DataTypeVal[String](5,
+    lenFixed          = false,
+    maxLen            = MaxShortTextLength,
+    deserializer      = b => new String(b.drop(2)),
+    serializer        = s => arrayShortLengthToByteArray(s.getBytes) ++ s.getBytes,
+    jsonifier         = Json.toJson(_),
+    extValidator      = _ => true)
 
-  sealed case class DataTypeVal(dataType: Int, lenFixed: Boolean, maxLen: Int) extends Val(dataType) { def *(n: Int): Int = n * dataType }
-
-  val PublicKey       = DataTypeVal(1, true, KeyLength)
-  val Address         = DataTypeVal(2, true, vsys.account.Address.AddressLength)
-  val Amount          = DataTypeVal(3, true, AmountLength)
-  val Int32           = DataTypeVal(4, true, 4)
-  val ShortText       = DataTypeVal(5, false, MaxShortTextLength)
-  val ContractAccount = DataTypeVal(6, true, vsys.account.ContractAccount.AddressLength)
-  val Account         = DataTypeVal(7, false, (vsys.account.Address.AddressLength).max(vsys.account.ContractAccount.AddressLength))
-  val TokenId         = DataTypeVal(8, true, vsys.account.ContractAccount.TokenAddressLength)
-  val Timestamp       = DataTypeVal(9, true, TimestampLength)
-  val Boolean         = DataTypeVal(10, true, 1)
-  val ShortBytes      = DataTypeVal(11, false, MaxShortBytesLength)
-  val Balance         = DataTypeVal(12, true, AmountLength)
-  val OpcBlock        = DataTypeVal(13, false, MaxOpcBlockLength)
-
-  def fromByte(b: Byte): Option[DataType.DataTypeVal] = Try(DataType(b).asInstanceOf[DataTypeVal]).toOption
+  val ContractAccount = DataTypeVal[vsys.account.ContractAccount](6,
+    lenFixed          = true,
+    maxLen            = vsys.account.ContractAccount.AddressLength,
+    deserializer      = vsys.account.ContractAccount.fromBytes(_).right.get,
+    serializer        = a => a.bytes.arr,
+    jsonifier         = a => Json.toJson(a.address),
+    extValidator      = vsys.account.ContractAccount.fromBytes(_).isRight)
+
+  val Account         = DataTypeVal[vsys.account.Account](7,
+    lenFixed          = false,
+    maxLen            = (vsys.account.Address.AddressLength).max(vsys.account.ContractAccount.AddressLength),
+    deserializer      = vsys.account.Account.fromBytes(_, 0).right.get._1,
+    serializer        = a => a.bytes.arr,
+    jsonifier         = a => Json.toJson(a.bytes.arr),
+    extValidator      = vsys.account.Account.fromBytes(_, 0).isRight)
```

Comment:
One thing a little bit confused here. `Account` type is not `allowed` as DataEntry passed in. It is a combination of `Address` and `ContractAccount` and which only used in function's input data check. But we can not pass an `Account` type `DataEntry` in previous logic. One simple solution we can use here is set `extValidator = _ => false`

Reply:
OK